### PR TITLE
feat: change implement for shards

### PR DESCRIPTION
**Pull request by hengfeiyang**
- feat: change implement for shards
- feat: add concurrent reads for multiple search
- feat: upgrade for old version <= v0.2.6

### CODE REVIEW

File: pkg/core/index.go [hunk 1]
```diff
@@ -250,8 +248,8 @@ func (index *Index) UpdateMetadataByShard(n int64) {
 	index.lock.Unlock()
 }
 
-func (index *Index) UpdateStatsBySecondShard(n, secondIndex int64) {
-	shard := index.shards[n]
+func (index *Index) UpdateStatsBySecondShard(id string, secondIndex int64) {
```
Comment (prabhatsharma) [inline pkg/core/index.go:251]: Can you add comments to every function saying what it is supposed to do.

File: pkg/metadata/index.go [hunk 2]
```diff
@@ -35,7 +36,16 @@ func (t *index) List(offset, limit int) ([]*meta.Index, error) {
 		idx := new(meta.Index)
 		err = json.Unmarshal(d, idx)
 		if err != nil {
-			return nil, err
+			if err.Error() == "expected { character for map value" {
```
Comment (danvixent) [inline pkg/metadata/index.go:39]: I think we should rather do if `strings.Contains` here.

File: pkg/ider/id.go [hunk 3]
```diff
@@ -16,27 +16,36 @@
...
+	return local.Generate()
+}
+
+func NewNode(id int) (*Node, error) {
+	node, err := snowflake.NewNode(int64(id % 1023))
+	return &Node{node: node}, err
+}
+
+func (n *Node) Generate() string {
+	return base62.Encode(n.node.Generate().Int64())
```
Comment (danvixent) [inline pkg/ider/id.go:50]: It's not clear why we're using base62 here.

Comment (hengfeiyang) [reply pkg/ider/id.go:50]: For shortly, we need generate a string for doc id, at beginning we use UUID, but it's very long. so we implement a new doc id.

File: pkg/metadata/index.go (hunk 2 above)
Comment (hengfeiyang) [reply pkg/metadata/index.go:39]: Yeah, do this for the exactly error. may be you got some error else?

File: pkg/core/index.go (hunk 1 above)
Comment (hengfeiyang) [reply pkg/core/index.go:251]: Okay

File: pkg/ider/id.go (hunk 3 above)
Comment (danvixent) [reply pkg/ider/id.go:50]: okay

**Review by prabhatsharma — APPROVED** [review]

File: pkg/bluge/search/search.go [hunk 4]
```diff
@@ -72,17 +92,31 @@ func MultiSearch(ctx context.Context, query *meta.ZincQuery, mappings *meta.Mapp
 		if err != nil {
 			return nil, err
 		}
+		if docList.sort == nil {
+			if req, ok := req.(*bluge.TopNSearch); ok {
+				docList.sort = req.SortOrder().Copy()
```
Comment (danvixent) [inline pkg/bluge/search/search.go:96]: Don't we sort for `bluge.AllMatches` as well?

Comment (hengfeiyang) [reply pkg/bluge/search/search.go:96]: In zinc, we just use topN search.

Comment (danvixent) [reply pkg/bluge/search/search.go:96]: okay

**Review by danvixent — APPROVED** [review]

**Pull request merged**